Feat: 정답 페이지 UI 구현

han-nun0107 opened the pull request:
## 🔥 Description (필수)

이 PR이 어떤 작업을 했는지 명확하게 작성해주세요.  
**최소 3줄 이상** 상세하게 작성 부탁드립니다.

- 질문 type에 대하여 수정 하였습니다. 기존 pull 받아왔던 컴포넌트의 경우 응시 화면 전용 컴포넌트라 이 컴포넌트에서 응시 결과 페이지 에서도 사용 가능하게 수정 작업 진행 했습니다.

- 코드의 경우 hooks에 들어가 있는 파일대로 코드가 변경 되었습니다. 사용자가 어떤 답을 했는지에 따라 텍스트의 색상이 바뀌고 사용자가 체크한 답안이 표시됩니다. 

- 오답과 정답의 상관 없이 문제에 대한 해답이 나오며 틀렸을 경우 붉은색으로 정답일 시 초록색으로 출력이 됩니다.

- 결과 페이지의 경우 공통된 Props를 상수로 묶어줘 코드 간결화를 시켰습니다.

---

## ⏰ Due (필수)

- 리뷰 마감 요청일: **2025.07.04까지 리뷰 부탁드립니다.**

---

## 🔄 변경 사항 (필수)

- [x] 기능 추가 (Feature)
- [ ] 버그 수정 (Bug Fix)
- [x] 리팩토링 (Refactor) — 기능 변경 없이 코드 개선
- [x] UI/스타일 수정 (UI/Style) — 기능 영향 없는 스타일, 디자인 변경
- [ ] 문서 수정 (Docs) — README, 문서화 등
- [ ] 테스트 코드 추가/수정 (Test)
- [ ] 배포 설정/환경 설정 변경 (Chore/Infra)
- [ ] 성능 최적화 (Performance)
- [ ] 기타 (Other)

---

## 🔍 Diff (필수)

| 파일명             | 역할                                        |
| ------------------ | ------------------------------------------- |
| `App.css` |  text-balance 기존에 tailwind에 있던 css라 네이밍이 겹쳐 layout이 깨지는 현상 해결하였습니다               |
| `CheckBoxType.tsx`       | 객관식 복수 선택 가능 문제 선택 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다.|               
| `OxType.tsx`    | OX 선택 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다.|
| `QuestionEmptyText.tsx`    | 빈칸 채우기 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다. |
| `QuestionResultSubBanner.tsx`    | min-w-[1920]을 넣어 overflow-x 가 생기는 현상을 min-w-[1920]을 제거하여 해결|
| `QuestionTextarea.tsx`    | 주관식 문제 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다 |
| `QuestionTitle.tsx`    | 기존 점수 부분과 문제 유형이 수직 수평 정렬이 안 되어 있는 것을 발견하여 해결|
| `RadioType.tsx`    | 객관식 단일 선택 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다. |
| `ReorderQuestion.tsx`    | DND 방식의 문제 컴포넌트이며 결과 페이지에도 사용 가능하게 코드 추가 하였습니다. |
| `ExamResultExplanation.tsx`    | 문제 설명 부분 결과 페이지에서 사용되는데 이 부분 공통으로 들어가져 컴포넌트로 분리 작업 후 추가 하였습니다.|
| `useExamOption.ts`    | 정답, 오답, 사용자가 클릭한 상태에 따라 텍스트 색상을 계산 해주는 유틸 함수를 추가 하였습니다.|
| `useExamValidation.ts`    | 전체 문제에 대한 정답 여부를 판별하는 유틸 함수를 추가하였습니다.|
| `examResultData.ts`    | api 명세서에 적혀있는 Mock 데이터를 그대로 가져와 적용 시켰습니다.|
| `ExamResult.tsx`    | 문제 결과 페이지이며 공통 된 Props의 경우 상수로 묶어 코드 간결화를 진행 하였습니다.|
| `examResult.ts`    | Mock 데이터의 타입을 지정 하였습니다. |

> ✅ 생성/수정된 파일은 "**파일명 : 파일 역할**" 형식으로 명확하게 작성해주세요.
---

### 📷 스크린샷 (UI 변경 시 필수)
![Image](https://github.com/user-attachments/assets/55472f01-181a-4bbc-93c9-cc8e0f5e6596)
---

## 💬 리뷰 요구사항 (선택)

> 리뷰어가 특히 집중해서 봐줬으면 하는 부분  
> 또는 고민되는 부분, 개선 아이디어가 필요하거나 질문하고 싶은 부분
> ex) 메서드 이름 `handleCheckOption` 이 적절한지 모르겠습니다. 더 좋은 이름 제안 부탁드립니다.
---

## 📄 기타 (선택)

> 추가적인 참고 사항이나 전달사항이 있으면 작성해주세요.
---
Feat: #43

han-nun0107 committed:
Feat: Mock 데이터 생성 및 Mock 데이터 타입 추가

feat: #43

han-nun0107 committed:
Feat: Mock 데이터 수정 및 데이터 타입 수정

feat: #43

han-nun0107 committed:
Feat: RadioType 결과 페이지에 맞게 수정

Feat: #43

han-nun0107 committed:
Feat: CheckBox 결과 페이지에 맞게 수정

Feat: #43

han-nun0107 committed:
Style: text-balance 코드 중복으로 인한 오류

text-balance 라는 tailwind 기본 코드가 있어 충돌
text-balance -> text-balance-space 라는 이름으로 변경

Style: #64

han-nun0107 committed:
Style: 문제 제목 부분 mb 추가

문제 부분과 문제 제목 부분의 간격이 있어 수정

Style: #43

han-nun0107 committed:
Style: overflow x 생기는 원인 수정

min x 1920 추가로 overflow x 원인 해결

style: #43

han-nun0107 committed:
Feat: CheckBox 결과 페이지에 맞게 수정

Feat: #43

han-nun0107 committed:
Feat: RadioType 결과 페이지에 맞게 수정

Feat: #43

han-nun0107 committed:
Feat: 결과 페이지에 맞게 컴포넌트 코드 수정 작업

- 수정하는 작업 중 중복되는 코드 리펙토링 작업도 같이 진행 했습니다.

Feat: #43

han-nun0107 committed:
Feat: 문제 결과 하단 부분 설명 컴포넌트 추가

- 기존에 컴포넌트를 분리 시키지 않아 코드가 길어지고 반복 되었던 것을 컴포넌트 분리를 하였습니다.

Feat: #43

han-nun0107 committed:
Feat: 시험 결과 페이지 전체 문제 정답 여부, 선택지 상태에 따른 텍스트 색상 변경 컴포넌트 추가

- 기존 반복되던 함수를 컴포넌트로 분리시켜 수정하기 쉽게 변경 하였습니다.

Feat: #43

han-nun0107 committed:
Feat: 시험 결과 페이지 컴포넌트

- 공통되는 Props의 경우 상수로 묶어 코드 간결화 시켰습니다.

Feat: #43

han-nun0107 committed:
Feat: 불 필요한 코드 제거

Feat: #43

devjaesung reviewed (dismissed):
수고하셨습니다 우수님!

nani6765 commented on src/components/examQuestions/CheckBoxType.tsx:
```diff
@@ -1,13 +1,16 @@
   disabled?: boolean
   student_answer?: string[]
   correct_answer?: string[]
+  explanation?: string
```
어떤 페이지에선 explanation값이 필수도 어떤 페이지에선 explanation값이 optional인가요?
explanation의 역할이 뭔가요?

nani6765 commented on src/hooks/examResult/useExamValidation.ts:
```diff
@@ -0,0 +1,19 @@
+  IS_WRONG_CHECK: boolean
+}
+
+export default function useExamValidation(
```
useHook이 아님

nani6765 commented on src/hooks/examResult/useExamOption.ts:
```diff
@@ -0,0 +1,33 @@
+/* 훅은 아닌데 유틸 폴더가 없어서 일단 임의로 훅 폴더에에 넣어둠 */
+
+export default function ExamOption(
```
1. useExamOption으로 이름이 변경되어야 함.
2. useHook이 아님

nani6765 commented on src/components/examQuestions/CheckBoxType.tsx:
```diff
@@ -39,52 +45,66 @@ export default function CheckBoxType({
+            student_answer,
+            selected,
+            correct_answer
+          )
```
인자가 3개 이상 넘어가면 구조분해 할당을 사용해보는게 좋을 것 같아요.

nani6765 commented on src/components/examQuestions/CheckBoxType.tsx:
```diff
@@ -39,52 +45,66 @@ export default function CheckBoxType({
-              <div className="absolute top-1/2 left-1/2 w-4 -translate-x-1/2 -translate-y-1/2 transform opacity-0 transition-opacity duration-200 peer-checked:opacity-100">
-                <img src="/src/assets/check.svg " className="h-full w-full" />
-              </div>
+          const checkboxId = `${question_Id}-${index}`
```
❤️

nani6765 commented on src/components/examQuestions/RadioType.tsx:
```diff
@@ -25,59 +30,68 @@ export default function RadioType({
+          )
+        })}
+      </div>
+      <ExamResultExplanation
```
`{is_result && <ExamResultExplanation/> }`

nani6765 commented on src/components/examQuestions/OxType.tsx:
```diff
@@ -34,111 +39,130 @@ export default function OxType({
+              if (IS_WRONG_OPTION) return 'bg-[#FDECEF]'
+              return 'bg-[#F2F3F5]'
+            }
+            return IS_CHECKED ? 'bg-[#EFE6FC]' : 'bg-[#F2F3F5]'
```
왜 내 리뷰 안들어줘요.

nani6765 replied:
함수 컴포넌트 return문 내부에 있을 필요는 없어욤~

nani6765 commented on src/components/examQuestions/OxType.tsx:
```diff
@@ -34,111 +39,130 @@ export default function OxType({
+                type="radio"
+                name={`question-${question_Id}`}
+                value={option}
+                checked={IS_CHECKED}
```
```
  const IS_CHECKED = is_result
    ? student_answer.includes(optionKey)
    : selected === optionKey
```

nani6765 replied:
checked가 되려면 
옵션키 = 옵션을 .으로 자른 첫번째 값
학생의 정답이 옵션키를 포함하고 있거나, `selected`의 값과 옵션키가 똑같은 경우

nani6765 replied:
문제를 다 푼 경우에는 수정 불가 의도

nani6765 commented on src/components/examQuestions/QuestionEmptyText.tsx:
```diff
@@ -36,44 +40,61 @@ export default function QuestionEmptyText({
+  const getTextColor = (value?: string, correct?: string) => {
+    const safeValue = value ?? ''
+    if (!is_result || safeValue.trim() === '') return 'text-black'
+    return safeValue === correct ? 'text-[#14C786]' : 'text-[#EC0037]'
```
`const getTextColor = ({value = "", correct} : {value?:string, cprrect?:string})`

nani6765 commented on src/components/examQuestions/QuestionTextarea.tsx:
```diff
@@ -1,52 +1,74 @@
+          value={value}
+          placeholder="20글자 이내로 입력해 주세요."
+          onChange={handleChange}
+          disabled={disabled}
```
disabeld={disabled || is_result}

nani6765 replied:
이 컴포넌트는 is_result일때 disabeld 동작을 안해도 되나요?

nani6765 commented on src/pages/ExamResult.tsx:
```diff
@@ -0,0 +1,125 @@
+import ReorderQuestion from '@/components/examQuestions/ReorderQuestion'
+import examResultData from '@/mock/examResultData'
+
+export default function ExamResult() {
```
```
const getContext = (type: Type) => {
  switch (type) {
    case 'multiple_choice_single':
      return (
        <RadioType
          {...COMMON_PROPS}
          options={question.options_json}
          question_Id={question.question_id}
        />
      )

    case 'multiple_choice_multiple':
      return (
        <CheckBoxType
          {...COMMON_PROPS}
          options={question.options_json}
          question_Id={question.question_id}
        />
      )
    case 'ox':
      return (
        <OxType
          {...COMMON_PROPS}
          options={question.options_json}
          question_Id={question.question_id}
        />
      )
    case 'ordering':
      return (
        <ReorderQuestion
          options_json={question.options_json}
          question_Id={question.question_id}
          {...COMMON_PROPS}
        />
      )

    case 'fill_in_blank':
      return (
        <QuestionTextarea
          {...COMMON_PROPS}
          question_Id={question.question_id}
          prompt={question.prompt as string}
        />
      )

    case 'short_answer':
      return (
        <QuestionEmptyText
          {...COMMON_PROPS}
          blank_count={question.blank_count ?? 0}
          prompt={question.prompt as string}
        />
      )

    default:
      return <div className="text-red-500">지원하지 않는 문제 타입입니다.</div>
  }
}

```

nani6765 commented on src/pages/ExamResult.tsx:
```diff
@@ -0,0 +1,125 @@
+                  score={question.point}
+                  type={question.type}
+                />
+                {questionComponent}
```
{getContext(question.type)}

han-nun0107 committed:
feat: Routes 부분 Rebase 충돌 해결

Feat: #43

han-nun0107 committed:
Refactor: 쪽지시험 오답 및 정답 설명 컴포넌트 수정

- 쪽지시험 오답 및 정답 설명 컴포넌트 수정
- is_result props로 내려주던 것을 제거하고 외부에서 처리하도록 수정

Refactor: #43

han-nun0107 committed:
Refactor: 쪽지시험 컴포넌트 중 ExamOption 사용 부분 수정

- 쪽지시험 컴포넌트 중 ExamOption 사용 부분 수정
- 쪽지시험 컴포넌트 중 ExamOption 부분 구조 분해로 사용 가능하게 수정

Refactor: #43

han-nun0107 committed:
Refactor: Routes 부분 Rebase 오류 해결

- switch case 문 map 밖으로 빼어 메모리 소비 줄임

Refactor: #43

han-nun0107 committed:
Refactor: 문제 타입 컴포넌트 수정 및 ExamOption 네이밍 수정

Refactor: #43

han-nun0107 committed:
Refactor: useExamValidation 수정으로 인한 코드 수정

- useExamValidation 구조분해로 수정 및 네이밍 수정

Refactor: #43

hi-tomato commented on src/hooks/examResult/useExamOption.ts:
`use/hooks` 파일로 분리할 수 있지만,
해당 파일은 react hook을 사용하지 않으니, 유틸함수 또는 헬퍼 함수라고 생각해봐도 되지 않을까 싶습니다!

han-nun0107 replied:
유틸 함수, 헬퍼 함수에 대해서 잘 모르고 있었는데 이번 기회에 찾아보면서 배울 수 있어 좋은 것 같습니다 !!
승준님 코드 리뷰 받으면서 많이 배워가는 것 같아요!
좋은 코드 리뷰 감사합니다!!

hi-tomato reviewed (approved):
고생하셨습니다 리뷰 확인 부탁드립니다 :)
